component/registry/registry.go

```diff
@@ -0,0 +1,63 @@
+package registry
+
+import (
+	"fmt"
+
+	"github.com/docker/docker/component"
+)
+
+// Registry stores Components so they can retrieved by other Components
+type Registry struct {
+	components map[string]component.Component
+}
+
+// Register adds a new component to the registry. If there is already a
+// component with the same ComponentType an error is returned.
+func (r *Registry) Register(c component.Component) {
+	r.components[c.Provides()] = c
+}
+
+// Get returns a Component by name that was previously registered
+func (r *Registry) Get(name string) (component.Component, error) {
+	component, ok := r.components[name]
+	if !ok {
+		return nil, fmt.Errorf("no component available for %q", name)
+	}
+	return component, nil
+}
+
+// All returns all the Components
+func (r *Registry) All() []component.Component {
+	comp := []component.Component{}
+	for _, component := range r.components {
+		comp = append(comp, component)
+	}
+	return comp
+}
+
+// ForEach runs a function on each component.Component
+func (r *Registry) ForEach(f func(component.Component) error) error {
```

Comment:
IMO one can use `for _, component := range r.All(){..}` instead of this `ForEach()`